src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -117,14 +118,23 @@ public Object execute(
     try {
       return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
     } catch (SQLException e) {
-      final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
-      final HostInfo currentHost = this.currentConnectionProvider.getCurrentHostInfo();
-      updateTopology(currentConnection, currentHost);
-      updateInternalConnections(currentConnection, currentHost);
+      if (isFailoverException(e)) {
+        final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+        final HostInfo currentHost = this.currentConnectionProvider.getCurrentHostInfo();
+        closeInternalConnection(this.readerConnection, currentConnection);
+        closeInternalConnection(this.writerConnection, currentConnection);
+        updateTopology(currentConnection, currentHost);
+        updateInternalConnections(currentConnection, currentHost);
+      }
       throw e;
     }
   }
 
+  private boolean isFailoverException(SQLException e) {
```

Comment:
Can we rename this method to something isCommunicationException()? I would like to avoid using "failover" term in this plugin.